Bluetooth: GATT: Add flag to indicate write command

Vudentz opened the pull request:
Application may need to handle the write differently depending on the write operation so this adds a flag called BT_GATT_WRITE_FLAG_CMD which can then be checked by the callback, for instance one can respond with BT_ATT_ERR_WRITE_REQ_REJECTED when that flag is not set which should indicate to the client to use write command instead.

Fixes #11206

Vudentz committed:
Bluetooth: GATT: Add flag to indicate write command

Application may need to handle the write differently depending on the
write operation so this adds a flag called BT_GATT_WRITE_FLAG_CMD which
can then be checked by the callback, for instance one can respond with
BT_ATT_ERR_WRITE_REQ_REJECTED when that flag is not set which should
indicate to the client to use write command instead.

Fixes #11206

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Fix comment related to write attribute

The comment was suggesting that write callback was reading and storing
buffer when in fact it should write to the attribute value.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -91,6 +91,13 @@ enum {
+
+	/** Attribute write command flag
+	 *
+	 * If set, indicates that write operation is a command which don't
```
doesn't

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -1204,9 +1205,12 @@ static u8_t write_cb(const struct bt_gatt_attr *attr, void *user_data)
 
-	/* Read attribute value and store in the buffer */
+	if (!data->op)
+		flags |= BT_GATT_WRITE_FLAG_CMD;
```
missing {}

Vudentz commented:
@jhedberg done with suggested changes

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -1204,9 +1205,13 @@ static u8_t write_cb(const struct bt_gatt_attr *attr, void *user_data)
 
-	/* Read attribute value and store in the buffer */
+	if (!data->op) {
+		flags |= BT_GATT_WRITE_FLAG_CMD;
```
How come data->op is 0 for BT_ATT_OP_WRITE_CMD and not, well, BT_ATT_OP_WRITE_CMD? Seems a bit strange. If there's a good reason for it there should at least be a short code comment explaining that 0 means BT_ATT_OP_WRITE_CMD.

Vudentz replied:
The intention was only to set it for requests which requires responses to avoid having extra checks for what type of request later on. Both write cmd and signed write cmd don't set a request opcode.

jhedberg reviewed (changes requested):
Thanks for doing the fixes I requested. One more question of something that's unclear to me.

Vudentz committed:
Bluetooth: ATT: Rename op parameter of att_write_rsp

This makes it clear that only request need to set an opcode since they
require a error response in case it fails.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@jhedberg anything preventing this to be merged?

Vudentz commented:
@carlescufi any chance to review this?

carlescufi commented on include/bluetooth/gatt.h:
```diff
@@ -91,6 +91,13 @@ enum {
+
+	/** Attribute write command flag
+	 *
+	 * If set, indicates that write operation is a command which doesn't
```
`If set, indicates that the write operation is a command (Write Without Response), which doesn't generate any response`

carlescufi reviewed (changes requested):
Thanks for this PR, just a comment on the API documentation
